From b4b8297f3031c1c112b25a30b2e40397370a3031 Mon Sep 17 00:00:00 2001
From: Jiuyang Liu <liu@jiuyang.me>
Date: Thu, 7 Oct 2021 12:03:37 +0800
Subject: [PATCH] Revert "remove mill support"

This reverts commit 07cb13a3b1d51b6aeb333824f2434bcf1b611f47.
For supporting mill build-from-source.
---
 .github/workflows/test.yml | 20 ++++++++-
 build.sc                   | 88 ++++++++++++++++++++++++++++++++++++++
 2 files changed, 107 insertions(+), 1 deletion(-)
 create mode 100644 build.sc

diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 411b137..be6a9be 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -80,13 +80,31 @@ jobs:
       - name: Check Formatting
         run: sbt scalafmtCheckAll
 
+  mill:
+    name: Mill Sanity Check
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        scala: [2.13.5, 2.12.13]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Cache
+        uses: coursier/cache-action@v5
+      - name: Setup Mill
+        uses: jodersky/setup-mill@v0.2.3
+        with:
+          mill-version: 0.9.9
+      - name: Mill sanity check
+        run: mill _[${{ matrix.scala }}].compile
+
   # Sentinel job to simplify how we specify which checks need to pass in branch
   # protection and in Mergify
   #
   # When adding new jobs, please add them to `needs` below
   all_tests_passed:
     name: "all tests passed"
-    needs: [test, doc]
+    needs: [test, doc, mill]
     runs-on: ubuntu-latest
     steps:
       - run: echo Success!
diff --git a/build.sc b/build.sc
new file mode 100644
index 0000000..466ab27
--- /dev/null
+++ b/build.sc
@@ -0,0 +1,88 @@
+import mill._
+import mill.scalalib._
+import mill.scalalib.publish._
+import mill.scalalib.scalafmt._
+import coursier.maven.MavenRepository
+
+object maltese extends mill.Cross[malteseCrossModule]("2.12.13", "2.13.5")
+
+class malteseCrossModule(val crossScalaVersion: String) extends CrossScalaModule with ScalafmtModule with PublishModule {
+  override def repositoriesTask = T.task {
+    super.repositoriesTask() ++ Seq(
+      MavenRepository("https://oss.sonatype.org/content/repositories/snapshots")
+    )
+  }
+
+  override def millSourcePath = super.millSourcePath / os.up
+
+  // 2.12.12 -> Array("2", "12", "12") -> "12" -> 12
+  private def majorVersion = crossScalaVersion.split('.')(1).toInt
+
+  def publishVersion = "0.5-SNAPSHOT"
+
+  private def javacCrossOptions = if(majorVersion == 12) Seq("-source", "1.8", "-target", "1.8") else Nil
+
+  override def scalacOptions = T {
+    super.scalacOptions() ++ Seq(
+      "-deprecation",
+      "-feature",
+      "-unchecked"
+    )
+  }
+
+  def treadleModule: Option[PublishModule] = None
+
+  def treadleIvyDeps = if (treadleModule.isEmpty) Agg(
+    ivy"edu.berkeley.cs::treadle:1.5-SNAPSHOT"
+  ) else Agg.empty[Dep]
+
+  def firrtlModule: Option[PublishModule] = None
+
+  def firrtlIvyDeps = if (treadleModule.isEmpty) Agg(
+    ivy"edu.berkeley.cs::firrtl:1.5-SNAPSHOT"
+  ) else Agg.empty[Dep]
+
+  override def javacOptions = T {
+    super.javacOptions() ++ javacCrossOptions
+  }
+
+  override def moduleDeps = super.moduleDeps ++ firrtlModule ++ treadleModule
+
+  override def ivyDeps = T {
+    super.ivyDeps() ++ Agg(
+      ivy"net.java.dev.jna:jna:5.4.0",
+      ivy"net.java.dev.jna:jna-platform:5.4.0",
+      ivy"com.github.com-github-javabdd:com.github.javabdd:1.0.1",
+    ) ++ firrtlIvyDeps ++ treadleIvyDeps
+  }
+
+  object test extends Tests {
+    override def millSourcePath = super.millSourcePath / os.up
+
+    override def ivyDeps = T {
+      Agg(
+        ivy"org.scalatest::scalatest:3.2.6"
+      )
+    }
+
+    override def testFramework = T {
+      "org.scalatest.tools.Framework"
+    }
+  }
+
+  def pomSettings = T {
+    PomSettings(
+      description = artifactName(),
+      organization = "edu.berkeley.cs",
+      url = "https://github.com/ucb-bar/maltese-smt",
+      licenses = Seq(License.`BSD-3-Clause`),
+      versionControl = VersionControl.github("ucb-bar", "maltese-smt"),
+      developers = Seq(
+        Developer("ekiwi", "Kevin Laeufer", "https://github.com/ekiwi/")
+      )
+    )
+  }
+
+  // make mill publish sbt compatible package
+  override def artifactName = "maltese-smt"
+}
